### PR TITLE
feat: support aten::matmul with different input shape

### PR DESCRIPTION
**Pull request by inocsin**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
support aten::matmul with diiferent input shape, like [1,32,64] * [64, 64]

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/conversion/converters/impl/matrix_multiply.cpp [hunk 1]
```diff
@@ -12,18 +12,45 @@ auto mm_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns().pattern(
...
+         shuffle->setReshapeDimensions(util::toDims(new_shape));
+         auto mm_layer = ctx->net->addMatrixMultiply(
+             *shuffle->getOutput(0), nvinfer1::MatrixOperation::kNONE, *other, nvinfer1::MatrixOperation::kNONE);
+         mm_layer->setName(util::node_info(n).c_str());
+         std::vector<int64_t> out_shape;
+         for (int i = 0; i < (int)in_shape.size() - 1; i++) {
+           out_shape.push_back(in_shape[i]);
+         }
+         out_shape.push_back(other_shape[other_shape.size() - 1]);
+         auto out_shuffle = ctx->net->addShuffle(*mm_layer->getOutput(0));
```
Comment (peri044) [inline core/conversion/converters/impl/matrix_multiply.cpp:41]: We can probably eliminate the use of output reshape layer here. TRT does automatic broadcasting if it seems a dimension size of `1`. [https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_matrix_multiply_layer.html#details](url).
I think maybe for the case of matmul of 3x2x3 and 3x3, the `3x3` matrix should be reshaped to `1x3x3` and TRT will broadcast it along the first dimension of matmul. ONNX code implementation [https://github.com/onnx/onnx-tensorrt/blob/fbebb144744b3be8defecd8478d74940056df305/builtin_op_importers.cpp#L2077](url)

**Review by peri044 — CHANGES_REQUESTED** [review]

**Review by peri044 — APPROVED** [review]

**Comment by peri044** [comment]
@narendasan LGTM. This is good to merge after your review

File: core/conversion/converters/impl/matrix_multiply.cpp [hunk 2]
```diff
@@ -12,18 +12,45 @@ auto mm_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns().pattern(
...
+           val = val * in_shape[i];
+         }
+         new_shape.push_back(val);
+         new_shape.push_back(in_shape[in_shape.size() - 1]);
+         shuffle->setReshapeDimensions(util::toDims(new_shape));
+         auto mm_layer = ctx->net->addMatrixMultiply(
+             *shuffle->getOutput(0), nvinfer1::MatrixOperation::kNONE, *other, nvinfer1::MatrixOperation::kNONE);
+         mm_layer->setName(util::node_info(n).c_str());
+         std::vector<int64_t> out_shape;
+         for (int i = 0; i < (int)in_shape.size() - 1; i++) {
```
[the remark this answers is not in the document]
Comment (inocsin) [reply core/conversion/converters/impl/matrix_multiply.cpp]: The code is outdated, you can review the lastest version

**Comment by borisfom** [comment]
Did you test dynamic dimensions? I tried merging your branch and ran on a use case with matmul of  tensor sizes [-1, 1, 512] by [512, 64]. TRTorchc crashed on serialization stage, complaining about that matmul: "Broadcast dimensions must be non-negative". 
I have tried fixing straightforward issue with reshape() being added, but that did not resolve the actual issue - still getting the same error: 

--- a/core/conversion/converters/impl/matrix_multiply.cpp
+++ b/core/conversion/converters/impl/matrix_multiply.cpp
@@ -35,7 +35,12 @@ auto mm_registrations TRTORCH_UNUSED = RegisterNodeConversionPatterns().pattern(
          for (int i = 0; i < min_dim; i++) {
            new_shape.push_back(old_shape[i]);
          }
+        for (int i = 0; i < new_shape.size(); i++) {
+          if (new_shape[i] == -1)
+            new_shape[i] = 0;
+        }
          shuffle->setReshapeDimensions(util::toDims(new_shape));
+        shuffle->setZeroIsPlaceholder(true);
          auto reshaped_tensor = shuffle->getOutput(0);
          auto new_self = in_shape.size() > other_shape.size() ? self : reshaped_tensor;
          auto new_other = other_shape.size() > in_shape.size() ? other : reshaped_tensor;

**Pull request closed without merging**